File: src/demos/ps/oo/PSLibSpell.java
package demos.ps.oo;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import demos.ps.oo.PSEffect.Effect;
import demos.ps.oo.PSEffect.EffectOutcome;
import demos.ps.oo.PSEffect.EffectPlace;
import demos.ps.oo.PSEffect.EffectTarget;
import demos.ps.oo.PSLibSound.PS1Sound;
import demos.ps.oo.PSMenu.Cancellable;

public class PSLibSpell {

	private static final Logger log = LogManager.getLogger(PSLibSpell.class);

		public interface Spell {
			int getMpCost();
			Effect getEffect();
		}
		
		public enum PS1Spell implements Spell {
			REST	(2, "Spell_Heal", Effect.CURE),
			CURE	(6, "Spell_Cure", Effect.CURE),
			W_REST	(2, "Spell_Heal", Effect.WCURE),
			W_CURE	(6, "Spell_Cure", Effect.WCURE),
			REVIVE	(12, "Spell_Rebirth", Effect.REVIVE),
			F_REVIVE(30, "Spell_FullRebirth", Effect.F_REVIVE),

			FLY		(8, "Spell_Fly", Effect.FLY),
			EXIT	(4, "Spell_Bypass", Effect.EXIT),
			TRAP	(2, "Spell_Untrap", Effect.TRAP),
			OPEN	(4, "Spell_Open", Effect.OPEN),

			ESCAPE	(2, "Spell_Bye", Effect.ESCAPE),
			CHAT	(2, "Spell_Chat", Effect.CHAT),
			TELE	(4, "Spell_Telepathy", Effect.TELE),

			ROPE	(4, "Spell_Rope", Effect.ROPE),
			ROPE_ALL(20, "Spell_RopeAll", Effect.ROPE_ALL),
			FEAR	(2, "Spell_Terror", Effect.FEAR),
			FEAR_ALL(12, "Spell_TerrorAll", Effect.FEAR_ALL),
			WALL	(6, "Spell_Waller", Effect.WALL),
			FORCE	(10, "Spell_Power", Effect.FORCE),
			PROT	(8, "Spell_Magic_waller", Effect.PROT),

			FIRE	(4, "Spell_Fire", Effect.FIRE),
			GI_FIRE	(14, "Spell_GiFire", Effect.GIFIRE),
			WIND	(12, "Spell_Wind", Effect.WIND),
			THUNDER (18, "Spell_Thunder", Effect.THUNDER), // Changed to 18
			
			LIGHT		(2, "Spell_Light", Effect.LIGHT),
			POWER_CURE 	(12, "Spell_PowerCure", Effect.CURE)
			;
			
			private int mpCost;
			private Effect effect;
			private String nameString;

			PS1Spell(int mpCost, String name, Effect effect) {
				this.mpCost = mpCost;
				this.nameString = name;
				this.effect = effect;
			}
			
			public Effect getEffect() {
				return this.effect;
			}
			public int getMpCost() {
				return this.mpCost;
			}

			@Override
			public String toString() {
				return PSGame.getString(nameString);
			}
 	};
		
		
 	// Called when the spell is select
 	public static PSEffect prepareSpell(Spell spell, PartyMember caster) {
 		
 		if(caster.mp < spell.getMpCost()) {
 			PSMenu.Stext(PSGame.getString("Magic_Not_Enough"));
 			return null;
 		}

 		PSEffect effect = new PSEffect(spell.getEffect());
 		effect.setUser(caster);
 		
 		if(spell.getEffect().getTarget() == EffectTarget.MEMBER || spell.getEffect().getTarget() == EffectTarget.ALIVE_MEMBER) {
			int partySel = 1;
			if(PSGame.getParty().partySize() > 1) {
				PSMenu.instance.push(PSMenu.instance.createPromptBox(130, 70, PSGame.getParty().listMembers(), true));
				partySel = PSMenu.instance.waitOpt(Cancellable.TRUE) + 1;
				PSMenu.instance.pop();
			}
			
			if(partySel == 0) {
				return null;
			}
			PartyMember target = PSGame.getParty().getMember(partySel-1);
			if(spell.getEffect().getTarget() == EffectTarget.ALIVE_MEMBER && target.getHp() <= 0) {
				PSMenu.Stext(PSGame.getString("Battle_Player_Dead", "<player>", target.getName()));
				return null;
			}
			
			effect.setTarget(target);
 		}
 		
 		if(spell == PS1Spell.REST || spell == PS1Spell.W_REST) {
 			effect.setValue(20);
 		} else if (spell == PS1Spell.CURE || spell == PS1Spell.W_CURE) {
 			effect.setValue(80);
 		} else if (spell == PS1Spell.POWER_CURE) {
 			effect.setValue(200);
 		}
 		
 		
 		return effect;
 	}
 	
 	// Called when spell is issued (after prepareSpell in World or at the proper turn in Battle)
 	public static EffectOutcome castSpell(Spell spell, PSEffect effect) {
 		
 		PartyMember p = effect.getUser();
 		
 		if(p.mp < spell.getMpCost()) { // Validate again for battle purposes
 			PSMenu.Stext(PSGame.getString("Magic_Not_Enough"));
 			return EffectOutcome.NONE;
 		}

 		if(	spell.getEffect() != Effect.FIRE &&	spell.getEffect() != Effect.WIND &&	spell.getEffect() != Effect.THUNDER) {
			PSGame.playSound(PS1Sound.SPELL);
 		}
 		
 		p.mp -= spell.getMpCost();

 		if(p.textBox != null) { // in battle
 			p.textBox.updateText(2, PSGame.getString("Stats_MP") + ":" + PSGame.format(p.getMp(), 4));
 		}
 		
 		EffectOutcome callEffect = effect.callEffect();
 		if(callEffect == EffectOutcome.FAIL) {
			PSMenu.Stext(PSGame.getString("Magic_Player_Fail", "<player>", effect.getUser().getName())); 			
 		}
 		return callEffect;
 	}
 	
 	
	
}

File: src/demos/ps/game/PSLibSpell.ts
/**
 * PSLibSpell - Phantasy Star Spell Library
 * Defines all spells, their costs, and effects
 */

import { Effect, EffectOutcome, EffectPlace, EffectTarget, PSEffect } from './PSEffect';
import { PartyMember } from './PartyMember';
import { PSGame } from '../PSGame';

// Spell interface
export interface Spell {
  getMpCost(): number;
  getEffect(): Effect;
}

// PS1 Spell enum
export enum PS1Spell {
  // Healing spells
  REST,
  CURE,
  W_REST,
  W_CURE,
  REVIVE,
  F_REVIVE,

  // Utility spells
  FLY,
  EXIT,
  TRAP,
  OPEN,

  // Communication spells
  ESCAPE,
  CHAT,
  TELE,

  // Control spells
  ROPE,
  ROPE_ALL,
  FEAR,
  FEAR_ALL,
  WALL,
  FORCE,
  PROT,

  // Attack spells
  FIRE,
  GI_FIRE,
  WIND,
  THUNDER,

  // Light spell
  LIGHT,
  POWER_CURE
}

export class PS1SpellHelper {
  private static readonly spellConfigs = new Map<PS1Spell, {
    mpCost: number,
    nameString: string,
    effect: Effect
  }>([
    // Healing spells
    [PS1Spell.REST, { mpCost: 2, nameString: "Spell_Heal", effect: Effect.CURE }],
    [PS1Spell.CURE, { mpCost: 6, nameString: "Spell_Cure", effect: Effect.CURE }],
    [PS1Spell.W_REST, { mpCost: 2, nameString: "Spell_Heal", effect: Effect.WCURE }],
    [PS1Spell.W_CURE, { mpCost: 6, nameString: "Spell_Cure", effect: Effect.WCURE }],
    [PS1Spell.REVIVE, { mpCost: 12, nameString: "Spell_Rebirth", effect: Effect.REVIVE }],
    [PS1Spell.F_REVIVE, { mpCost: 30, nameString: "Spell_FullRebirth", effect: Effect.F_REVIVE }],

    // Utility spells
    [PS1Spell.FLY, { mpCost: 8, nameString: "Spell_Fly", effect: Effect.FLY }],
    [PS1Spell.EXIT, { mpCost: 4, nameString: "Spell_Bypass", effect: Effect.EXIT }],
    [PS1Spell.TRAP, { mpCost: 2, nameString: "Spell_Untrap", effect: Effect.TRAP }],
    [PS1Spell.OPEN, { mpCost: 4, nameString: "Spell_Open", effect: Effect.OPEN }],

    // Communication spells
    [PS1Spell.ESCAPE, { mpCost: 2, nameString: "Spell_Bye", effect: Effect.ESCAPE }],
    [PS1Spell.CHAT, { mpCost: 2, nameString: "Spell_Chat", effect: Effect.CHAT }],
    [PS1Spell.TELE, { mpCost: 4, nameString: "Spell_Telepathy", effect: Effect.TELE }],

    // Control spells
    [PS1Spell.ROPE, { mpCost: 4, nameString: "Spell_Rope", effect: Effect.ROPE }],
    [PS1Spell.ROPE_ALL, { mpCost: 20, nameString: "Spell_RopeAll", effect: Effect.ROPE_ALL }],
    [PS1Spell.FEAR, { mpCost: 2, nameString: "Spell_Terror", effect: Effect.FEAR }],
    [PS1Spell.FEAR_ALL, { mpCost: 12, nameString: "Spell_TerrorAll", effect: Effect.FEAR_ALL }],
    [PS1Spell.WALL, { mpCost: 6, nameString: "Spell_Waller", effect: Effect.WALL }],
    [PS1Spell.FORCE, { mpCost: 10, nameString: "Spell_Power", effect: Effect.FORCE }],
    [PS1Spell.PROT, { mpCost: 8, nameString: "Spell_Magic_waller", effect: Effect.PROT }],

    // Attack spells
    [PS1Spell.FIRE, { mpCost: 4, nameString: "Spell_Fire", effect: Effect.FIRE }],
    [PS1Spell.GI_FIRE, { mpCost: 14, nameString: "Spell_GiFire", effect: Effect.GIFIRE }],
    [PS1Spell.WIND, { mpCost: 12, nameString: "Spell_Wind", effect: Effect.WIND }],
    [PS1Spell.THUNDER, { mpCost: 18, nameString: "Spell_Thunder", effect: Effect.THUNDER }],

    // Light spell
    [PS1Spell.LIGHT, { mpCost: 2, nameString: "Spell_Light", effect: Effect.LIGHT }],
    [PS1Spell.POWER_CURE, { mpCost: 12, nameString: "Spell_PowerCure", effect: Effect.CURE }]
  ]);

  public static getMpCost(spell: PS1Spell): number {
    return this.spellConfigs.get(spell)?.mpCost || 0;
  }

  public static getEffect(spell: PS1Spell): Effect {
    return this.spellConfigs.get(spell)?.effect || Effect.NONE;
  }

  public static toString(spell: PS1Spell): string {
    const nameString = this.spellConfigs.get(spell)?.nameString;
    if (nameString) {
      return PSGame.getString(nameString);
    }
    return spell.toString();
  }
}

// PS1SpellImpl class implements Spell interface for each spell
export class PS1SpellImpl implements Spell {
  private spell: PS1Spell;

  constructor(spell: PS1Spell) {
    this.spell = spell;
  }

  public getMpCost(): number {
    return PS1SpellHelper.getMpCost(this.spell);
  }

  public getEffect(): Effect {
    return PS1SpellHelper.getEffect(this.spell);
  }

  public toString(): string {
    return PS1SpellHelper.toString(this.spell);
  }

  public getSpell(): PS1Spell {
    return this.spell;
  }
}

// Factory class for creating spell instances
export class SpellFactory {
  public static createSpell(spell: PS1Spell): Spell {
    return new PS1SpellImpl(spell);
  }
}

// Spell preparation and casting functions (simplified for now)
// Note: These would need full PSMenu implementation

export class PSLibSpell {
  /**
   * Prepare spell for casting - simplified version of Java prepareSpell
   */
  public static prepareSpell(spell: Spell, caster: PartyMember): PSEffect | null {
    // Implementation would require full PSEffect, PSMenu, and PartyMember classes
    console.warn('prepareSpell not yet fully implemented - requires PSEffect, PSMenu, and PartyMember');
    return null;
  }

  /**
   * Cast prepared spell - simplified version of Java castSpell
   */
  public static castSpell(spell: Spell, effect: PSEffect): EffectOutcome {
    // Implementation would require full PSEffect, PSMenu, and PartyMember classes
    console.warn('castSpell not yet fully implemented - requires PSEffect, PSMenu, and PartyMember');
    return EffectOutcome.NONE;
  }
}